src/rectangle_iterator.cpp

```diff
@@ -0,0 +1,130 @@
+#include "grid_map_proc/rectangle_iterator.h"
+
+#include <ros/console.h>
+
+#include <grid_map_core/GridMapMath.hpp>
+#include <grid_map_core/SubmapGeometry.hpp>
+
+#include <algorithm>
+#include <limits>
+#include <stdexcept>
+#include <string>
+
+namespace grid_map_proc
+{
+RectangleIterator::RectangleIterator(const grid_map::GridMap& map, const grid_map::Polygon& rectangle_polygon)
+  : polygon_{ rectangle_polygon }, past_end_{ false }
+{
+  const std::vector<grid_map::Position>& vertices = polygon_.getVertices();
+  if (vertices.size() != 4U)
+  {
+    std::string error{ "RectangleIterator needs 4 point polygone, got size " + std::to_string(vertices.size()) };
+    ROS_FATAL_STREAM(error);
+    throw std::domain_error(error);
+  }
+
+  const grid_map::Position& front_point = vertices.front();
+  double min_x = front_point.x();
+  double max_x = front_point.x();
+
+  double min_y = front_point.y();
+  double max_y = front_point.y();
+
+  for (uint8_t i = 1; i < 4U; ++i)
+  {
+    const grid_map::Position& point = vertices.at(i);
+    min_x = std::min(min_x, point.x());
+    max_x = std::max(max_x, point.x());
+
+    min_y = std::min(min_y, point.y());
+    max_y = std::max(max_y, point.y());
+  }
+
+  const auto get_index = [&map](const double x, const double y) -> grid_map::Index {
+    grid_map::Position position{ x, y };
+    position = map.getClosestPositionInMap(position);
+
+    grid_map::Index index;
+    map.getIndex(position, index);
+
+    return index;
+  };
+
+  start_index_ = get_index(max_x, max_y);
+  end_index_ = get_index(min_x, min_y);
```

Comment:
Why is end min_x/y if you increment in nextStep?

Reply:
Position coordinates and index directions are flipped:
https://github.com/ANYbotics/grid_map/blob/master/grid_map_core/doc/grid_map_conventions.pdf

I added a comment